File: projects/10-multi-runtime-api-benchmark/lambdas/python/src/utils/dynamodb.py
import os
import time
import uuid
from typing import Optional, List
from decimal import Decimal
import boto3
from botocore.exceptions import ClientError
from aws_lambda_powertools import Logger

from ..models.item import Item, ItemCreate, ItemUpdate

logger = Logger()


class DynamoDBClient:
    """DynamoDB client for item operations"""

    def __init__(self):
        self.table_name = os.environ.get('TABLE_NAME', 'dev-benchmark-items')
        self.dynamodb = boto3.resource('dynamodb')
        self.table = self.dynamodb.Table(self.table_name)
        logger.info(f"DynamoDB client initialized for table: {self.table_name}")

    @staticmethod
    def _current_timestamp() -> int:
        """Get current timestamp in milliseconds"""
        return int(time.time() * 1000)

    @staticmethod
    def _decimal_to_float(obj):
        """Convert Decimal to float for JSON serialization"""
        if isinstance(obj, Decimal):
            return float(obj)
        raise TypeError

    def create_item(self, item_data: ItemCreate) -> Item:
        """Create a new item in DynamoDB"""
        item_id = str(uuid.uuid4())
        current_time = self._current_timestamp()

        item = {
            'id': item_id,
            'name': item_data.name,
            'description': item_data.description or '',
            'price': Decimal(str(item_data.price)),
            'created_at': current_time,
            'updated_at': current_time,
        }

        try:
            self.table.put_item(Item=item)
            logger.info(f"Created item: {item_id}")
            return Item(**item)
        except ClientError as e:
            logger.exception("Error creating item")
            raise

    def get_item(self, item_id: str) -> Optional[Item]:
        """Get an item by ID"""
        try:
            response = self.table.get_item(Key={'id': item_id})
            if 'Item' in response:
                logger.info(f"Retrieved item: {item_id}")
                return Item(**response['Item'])
            logger.warning(f"Item not found: {item_id}")
            return None
        except ClientError as e:
            logger.error(f"Error getting item {item_id}: {e}")
            raise

    def update_item(self, item_id: str, item_data: ItemUpdate) -> Optional[Item]:
        """Update an existing item"""
        # First check if item exists
        existing_item = self.get_item(item_id)
        if not existing_item:
            return None

        # Build update expression
        update_parts = []
        expression_values = {}
        expression_names = {}

        if item_data.name is not None:
            update_parts.append('#n = :name')
            expression_values[':name'] = item_data.name
            expression_names['#n'] = 'name'

        if item_data.description is not None:
            update_parts.append('description = :description')
            expression_values[':description'] = item_data.description

        if item_data.price is not None:
            update_parts.append('price = :price')
            expression_values[':price'] = Decimal(str(item_data.price))

        # Always update timestamp
        update_parts.append('updated_at = :updated_at')
        expression_values[':updated_at'] = self._current_timestamp()

        update_expression = 'SET ' + ', '.join(update_parts)

        try:
            response = self.table.update_item(
                Key={'id': item_id},
                UpdateExpression=update_expression,
                ExpressionAttributeValues=expression_values,
                ExpressionAttributeNames=expression_names if expression_names else None,
                ReturnValues='ALL_NEW'
            )
            logger.info(f"Updated item: {item_id}")
            return Item(**response['Attributes'])
        except ClientError as e:
            logger.error(f"Error updating item {item_id}: {e}")
            raise

    def delete_item(self, item_id: str) -> bool:
        """Delete an item"""
        try:
            # Check if item exists first
            existing_item = self.get_item(item_id)
            if not existing_item:
                return False

            self.table.delete_item(Key={'id': item_id})
            logger.info(f"Deleted item: {item_id}")
            return True
        except ClientError as e:
            logger.error(f"Error deleting item {item_id}: {e}")
            raise

    def list_items(
        self,
        limit: int = 100,
        exclusive_start_key: Optional[dict] = None
    ) -> tuple[List[Item], Optional[dict]]:
        """
        List items with optional pagination support

        Args:
            limit: Maximum number of items to return (default: 100)
            exclusive_start_key: Optional key to start scanning from (for pagination)

        Returns:
            Tuple of (items, last_evaluated_key) where last_evaluated_key is None if no more items
        """
        try:
            actual_limit = limit if limit > 0 else 100

            scan_kwargs = {'Limit': actual_limit}
            if exclusive_start_key:
                scan_kwargs['ExclusiveStartKey'] = exclusive_start_key

            response = self.table.scan(**scan_kwargs)
            items = [Item(**item) for item in response.get('Items', [])]
            last_evaluated_key = response.get('LastEvaluatedKey')

            has_more = last_evaluated_key is not None
            logger.info(f"Listed {len(items)} items (hasMore: {has_more})")

            return items, last_evaluated_key
        except ClientError as e:
            logger.error(f"Error listing items: {e}")
            raise

File: projects/10-multi-runtime-api-benchmark/lambdas/typescript/src/utils/dynamodb.ts
import {
  DynamoDBClient,
  GetItemCommand,
  PutItemCommand,
  UpdateItemCommand,
  DeleteItemCommand,
  ScanCommand,
} from '@aws-sdk/client-dynamodb';
import { marshall, unmarshall } from '@aws-sdk/util-dynamodb';
import { v4 as uuidv4 } from 'uuid';
import { Item, ItemCreate, ItemUpdate } from '../models/item';

export class DynamoDBService {
  private client: DynamoDBClient;
  private tableName: string;

  constructor() {
    this.client = new DynamoDBClient({});
    this.tableName = process.env.TABLE_NAME || 'dev-benchmark-items';
    console.log(`DynamoDB service initialized for table: ${this.tableName}`);
  }

  private currentTimestamp(): number {
    return Date.now();
  }

  async createItem(itemData: ItemCreate): Promise<Item> {
    const item: Item = {
      id: uuidv4(),
      name: itemData.name,
      description: itemData.description || '',
      price: itemData.price,
      created_at: this.currentTimestamp(),
      updated_at: this.currentTimestamp(),
    };

    try {
      await this.client.send(
        new PutItemCommand({
          TableName: this.tableName,
          Item: marshall(item),
        })
      );
      console.log(`Created item: ${item.id}`);
      return item;
    } catch (error) {
      console.error(`Error creating item:`, error);
      throw error;
    }
  }

  async getItem(itemId: string): Promise<Item | null> {
    try {
      const response = await this.client.send(
        new GetItemCommand({
          TableName: this.tableName,
          Key: marshall({ id: itemId }),
        })
      );

      if (!response.Item) {
        console.warn(`Item not found: ${itemId}`);
        return null;
      }

      const item = unmarshall(response.Item) as Item;
      console.log(`Retrieved item: ${itemId}`);
      return item;
    } catch (error) {
      console.error(`Error getting item ${itemId}:`, error);
      throw error;
    }
  }

  async updateItem(itemId: string, itemData: ItemUpdate): Promise<Item | null> {
    // First check if item exists
    const existingItem = await this.getItem(itemId);
    if (!existingItem) {
      return null;
    }

    // Build update expression
    const updateParts: string[] = [];
    const expressionValues: Record<string, any> = {};
    const expressionNames: Record<string, string> = {};

    if (itemData.name !== undefined) {
      updateParts.push('#n = :name');
      expressionValues[':name'] = itemData.name;
      expressionNames['#n'] = 'name';
    }

    if (itemData.description !== undefined) {
      updateParts.push('description = :description');
      expressionValues[':description'] = itemData.description;
    }

    if (itemData.price !== undefined) {
      updateParts.push('price = :price');
      expressionValues[':price'] = itemData.price;
    }

    // Always update timestamp
    updateParts.push('updated_at = :updated_at');
    expressionValues[':updated_at'] = this.currentTimestamp();

    const updateExpression = 'SET ' + updateParts.join(', ');

    try {
      const response = await this.client.send(
        new UpdateItemCommand({
          TableName: this.tableName,
          Key: marshall({ id: itemId }),
          UpdateExpression: updateExpression,
          ExpressionAttributeValues: marshall(expressionValues),
          ExpressionAttributeNames: Object.keys(expressionNames).length > 0 ? expressionNames : undefined,
          ReturnValues: 'ALL_NEW',
        })
      );

      if (!response.Attributes) {
        return null;
      }

      const item = unmarshall(response.Attributes) as Item;
      console.log(`Updated item: ${itemId}`);
      return item;
    } catch (error) {
      console.error(`Error updating item ${itemId}:`, error);
      throw error;
    }
  }

  async deleteItem(itemId: string): Promise<boolean> {
    try {
      // Check if item exists first
      const existingItem = await this.getItem(itemId);
      if (!existingItem) {
        return false;
      }

      await this.client.send(
        new DeleteItemCommand({
          TableName: this.tableName,
          Key: marshall({ id: itemId }),
        })
      );

      console.log(`Deleted item: ${itemId}`);
      return true;
    } catch (error) {
      console.error(`Error deleting item ${itemId}:`, error);
      throw error;
    }
  }

  async listItems(limit: number = 100): Promise<Item[]> {
    try {
      const response = await this.client.send(
        new ScanCommand({
          TableName: this.tableName,
          Limit: limit,
        })
      );

      const items = (response.Items || []).map((item) => unmarshall(item) as Item);
      console.log(`Listed ${items.length} items`);
      return items;
    } catch (error) {
      console.error(`Error listing items:`, error);
      throw error;
    }
  }
}
